### ISSUE TITLE
Check that R packages required by the lesson are installed

### ISSUE THREAD

**Issue by fmichonneau** [swcarpentry/styles#82]
We need to check that all the packages required by a lesson, and this check should probably live in this repository so the code is consistent across lessons that use R packages.

The problem is that, a priori, this repository doesn't know which package a lesson will need. I can think of a couple of approaches:

1. We hard code in this repository all packages used across lessons based on R. Probably the easiest, but not very practical, as adding a new package dependency in a lesson requires a pull request here, and merging the changes back in the lesson repository.

2. We create a file within each repository that lists the packages the lesson uses (à la DESCRIPTION). A script in the styles repository would parse this file and check that the packages are available.

3. We use a per lesson approach and let maintainers make the appropriate modifications to the code in their lessons to make sure that the packages their lessons rely on are available.

I think 2 is my preferred option. Any thoughts or better solution @naupaka, @noamross, @gvwilson?

**Comment by naupaka** [comment]
2 seems the cleanest but requires contributors to be aware that they need to change that PACKAGES file as well if their code adds a new package. 

3 seems like the lowest barrier approach, but would leave lots of `if(!require()) install.packages()` statements all over the place, which isn't the cleanest approach either, although I would be curious what people think is the best general approach to this problem when sharing code among collaborators. 

I think 1 probably is too much hassle and could lead to problems depending on when changes are merged in the two repos.

**Comment by noamross** [comment]
The following bit of code will scan the directory for required packages and install those that are missing, avoiding the need to manually track the necessary packages or put `!require()` statements everywhere.  It can go at the start of `bin/generate_md_episodes.R`.

We could use the **checkpoint** package to fix the versions of packages used, too, but we probably _want_ to see when package updates break the lessons, as students will generally be installing the latest versions of packages.

```
if (require("checkpoint")) {
  required_pkgs <-
    checkpoint:::projectScanPackages(verbose=FALSE, use.knitr = TRUE)$pkgs
} else {
  stop("The checkpoint package is required to build the lessons.")
}

install.packages(
  required_pkgs[!(required_pkgs %in% rownames(installed.packages()))]
)
```

**Comment by fmichonneau** [comment]
Thanks @noamross, that looks good! I implemented your proposed solution in #83 .
